Raise code error when git command fails

gridbugs opened the pull request:
This is intended to help developers faster understand what's happening when dune invokes git and git fails with an error.

rgrinberg commented on src/dune_pkg/rev_store.ml:
```diff
@@ -100,26 +100,61 @@ let make_stderr () = Process.Io.make_stderr ~output_on_success:Swallow ~output_l
+  let stderr_path = Temp.create File ~prefix:"dune" ~suffix:"output" in
+  let stderr_to = Process.Io.file stderr_path Out in
+  let out =
+    Fiber.with_error_handler
```
Doesn't this hide the original exception?

What about just changing `failure_mode` to just return the status and we can fail ourselves with a proper error.

gridbugs commented:
> Isn't the invocation already logged to `_build/log`? This is what I check when there's some problem with subprocesses.

I think for the sake of ux it should be a `Code_error` though. Currently the only output is the error message from git with no context.

gridbugs committed:
pkg: code error when git/tar command fails

It's a programming error (not a user error) if package management code
invokes git or tar in a way which unexpectedly returns a non-zero value.
This change raises a Code_error in such cases.

Signed-off-by: Stephen Sherratt <stephen@sherra.tt>

gridbugs commented:
I rewrote this to set `failure_mode` to `Return`

rgrinberg committed:
_

Signed-off-by: Rudi Grinberg <me@rgrinberg.com>

rgrinberg commented:
Merging this after some simplifications.

Note that `==` is physical equality and while it works as you'd expect on integers, it's unlikely what you want. I removed it.
